README.md
# NAME

URL::Normalize - Normalize/optimize URLs.

# VERSION

Version 0.44

# SYNOPSIS

    use URL::Normalize;

    my $normalizer = URL::Normalize->new( 'http://www.example.com/display?lang=en&article=fred' );

    # Normalize the URL
    $normalizer->make_canonical;
    $normalizer->remove_directory_index;
    $normalizer->remove_empty_query;

    # Get the normalized version back
    my $url = $normalizer->url;

# DESCRIPTION

When writing a web crawler, for example, it's always very costly to check if a
URL has been fetched/seen when you have millions or billions of URLs in a
database.

This module can help you create a unique "ID" of a URL, which you can use as a
key in a key/value-store; the key is the normalized URL, whereas all the URLs
that refers to the normalized URL are part of the value (normally an array or
hash);

    'http://www.example.com/' = {
        'http://www.example.com:80/'        => 1,
        'http://www.example.com/index.html' => 1,
        'http://www.example.com/?'          => 1,
    }

Above, all the URLs inside the hash normalizes to the key if you run these
methods:

- `make_canonical`
- `remove_directory_index`
- `remove_empty_query`

This is NOT a perfect solution.

If you normalize a URL using all the methods in this module, there is a high
probability that the URL will stop "working." This is merely a helper module
for those of you who wants to either normalize a URL using only a few of the
safer methods, and/or for those of you who wants to generate a possibly unique
"ID" from any given URL.

# CONSTRUCTORS

## new( $url )

Constructs a new URL::Normalize object:

    my $normalizer = URL::Normalize->new( 'http://www.example.com/some/path' );

You can also send in just the path:

    my $normalizer = URL::Normalize->new( '/some/path' );

The latter is NOT recommended, though, and hasn't been tested properly. You
should always give URL::Normalize an absolute URL by using [URI](https://metacpan.org/pod/URI)'s `new_abs`.

# METHODS

## url

Get the current URL, preferably after you have run one or more of the
normalization methods.

## URI

Returns a [URI](https://metacpan.org/pod/URI) representation of the current URL.

## make\_canonical

Just a shortcut for URI::URL->new->canonical->as\_string, and involves the
following steps (at least):

- Converts the scheme and host to lower case.
- Capitalizes letters in escape sequences.
- Decodes percent-encoded octets of unreserved characters.
- Removes the default port (port 80 for http).

Example:

    my $normalizer = URL::Normalize->new(
        url => 'HTTP://www.example.com:80/%7Eusername/',
    );

    $normalizer->make_canonical;

    print $normalizer->url; # http://www.example.com/~username/

## remove\_dot\_segments

The `.`, `..` and `...` segments will be removed and "folded" (or
"flattened", if you prefer) from the URL.

This method does NOT follow the algorithm described in [RFC 3986: Uniform
Resource Indentifier](http://tools.ietf.org/html/rfc3986), but rather flattens
each path segment.

Also keep in mind that this method doesn't (because it can't) account for
symbolic links on the server side.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/../a/b/../c/./d.html',
    );

    $normalizer->remove_dot_segments;

    print $normalizer->url; # http://www.example.com/a/c/d.html

## remove\_directory\_index

Removes well-known directory indexes, eg. `index.html`, `default.asp` etc.
This method is case-insensitive.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/index.cgi?foo=/',
    );

    $normalizer->remove_directory_index;

    print $normalizer->url; # http://www.example.com/?foo=/

The default regular expressions for matching a directory index are:

- `default\.aspx?`
- `default\.html\.aspx?`
- `default\.s?html?`
- `home\.s?html?`
- `index\.cgi`
- `index\.html\.aspx?`
- `index\.html\.php`
- `index\.jsp`
- `index\.php\d?`
- `index\.pl`
- `index\.s?html?`
- `welcome\.s?html?`

You can override these by sending in your own list of regular expressions
when creating the URL::Normalizer object:

    my $normalizer = URL::Normalize->new(
        url               => 'http://www.example.com/index.cgi?foo=/',
        dir_index_regexps => [ 'MyDirIndex\.html' ], # etc.
    );

You can also choose to add regular expressions after the URL::Normalize
object has been created:

    my $normalizer = URL::Normalize->new(
        url               => 'http://www.example.com/index.cgi?foo=/',
        dir_index_regexps => [ 'MyDirIndex\.html' ], # etc.
    );

    # ...

    $normalizer->add_directory_index_regexp( 'MyDirIndex\.html' );

## sort\_query\_parameters

Sorts the URL's query parameters alphabetically.

Uppercased parameters will be lowercased DURING sorting, but the parameters
will be in the original case AFTER sorting. If there are multiple values for
one parameter, the key/value-pairs will be sorted as well.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/?b=2&c=3&a=0&A=1',
    );

    $normalizer->sort_query_parameters;

    print $normalizer->url; # http://www.example.com/?a=0&A=1&b=2&c=3

## remove\_duplicate\_query\_parameters

Removes duplicate query parameters, i.e. where the key/value combination is
identical with another key/value combination.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/?a=1&a=2&b=4&a=1&c=4',
    );

    $normalizer->remove_duplicate_query_parameters;

    print $normalizer->url; # http://www.example.com/?a=1&a=2&b=3&c=4

## remove\_empty\_query\_parameters

Removes empty query parameters, i.e. where there are keys with no value. This
only removes BLANK values, not values considered to be no value, like zero (0).

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/?a=1&b=&c=3',
    );

    $normalizer->remove_empty_query_parameters;

    print $normalizer->url; # http://www.example.com/?a=1&c=3

## remove\_empty\_query

Removes empty query from the URL.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/foo?',
    );

    $normalizer->remove_empty_query;

    print $Normalize->url; # http://www.example.com/foo

## remove\_fragment

Removes the fragment from the URL, but only if seems like they are at the end
of the URL.

For example `http://www.example.com/#foo` will be translated to
`http://www.example.com/`, but `http://www.example.com/#foo/bar` will stay
the same.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/bar.html#section1',
    );

    $normalizer->remove_fragment;

    print $normalizer->url; # http://www.example.com/bar.html

## remove\_fragments

Like `remove_fragment`, but removes EVERYTHING after a `#`.

## remove\_duplicate\_slashes

Remove duplicate slashes from the URL.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/foo//bar.html',
    );

    $normalizer->remove_duplicate_slashes;

    print $normalizer->url; # http://www.example.com/foo/bar.html

## remove\_query\_parameter

Convenience method for removing a specific parameter from the URL. If
the parameter is mentioned multiple times (?a=1&a=2), all occurences
will be removed.

## remove\_query\_parameters

Convenience method for removing multiple parameters from the URL. If the
parameters are mentioned multiple times (?a=1&a=2), all occurences will be
removed.

# SEE ALSO

- [URI::Normalize](https://metacpan.org/pod/URI%3A%3ANormalize)
- [URI](https://metacpan.org/pod/URI)
- [URI::URL](https://metacpan.org/pod/URI%3A%3AURL)
- [URI::QueryParam](https://metacpan.org/pod/URI%3A%3AQueryParam)
- [RFC 3986: Uniform Resource Indentifier](http://tools.ietf.org/html/rfc3986)
- [Wikipedia: URL normalization](http://en.wikipedia.org/wiki/URL_normalization)

# AUTHOR

Tore Aursand, `<toreau at gmail.com>`

# BUGS

Please report any bugs or feature requests to the web interface at [https://github.com/toreau/URL-Normalize/issues](https://github.com/toreau/URL-Normalize/issues)

# SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc URL::Normalize

You can also look for information at:

- AnnoCPAN: Annotated CPAN documentation

    [http://annocpan.org/dist/URL-Normalize](http://annocpan.org/dist/URL-Normalize)

- CPAN Ratings

    [http://cpanratings.perl.org/d/URL-Normalize](http://cpanratings.perl.org/d/URL-Normalize)

- Search CPAN

    [http://search.cpan.org/dist/URL-Normalize/](http://search.cpan.org/dist/URL-Normalize/)

# LICENSE AND COPYRIGHT

The MIT License (MIT)

Copyright (c) 2012-2021 Tore Aursand

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


Changes
Revision history for URL::Normalize

0.44    2023-10-10
	- Switched from Dist::Zilla to Minilla.
        - Optimized removing directory indexes slightly.

0.43    2021-08-23
        - Added missing documentation/Changes information. This is what happens
          when it's years between each CPAN release. :(

0.41    2021-08-23
        - Fixed wrong release date in previous release.

0.40    2021-08-23
        - Added more default directory index regular epxressions.
        - Removed handling of social query parameters; it was always too
          specific/limited, and shouldn't be in a module like this one.
        - Minor fixes/improvements to the documentation.

0.39    2018-06-12
        - Remove unneeded test script, which should fix some failing tests.

0.38    2018-05-30
        - Canonicalise URLs when creating the object. Handles some issues that
          URI::Normalize solved better than this module.

0.37    2018-05-30
        - Rewrote 'remove_dot_segments' to handle fragments better. This fixes
          issue #3. Improved tests accordingly. (thanks cyga)
        - Added dzil [MetaJSON]. (thanks manwar)
        - Upped versions on dependencies.

0.36    2017-12-27
        - Added 'remove_query_parameter' and 'remove_empty_query_parameters'.
        - Deprecated 'get_url'; 'url' should be used instead.
        - Upped dependencies' version requirements.
        - Improved tests.
        - Improved documentation.

0.35    2016-07-07
        - Corrected the version number in the documentation.

0.34    2016-07-07
        - MAJOR: Directory indexes are NO LONGER case-sensitive, so 'Index.html'
                 and 'index.html' are treated the same.
        - Added more directory indexes to the default directory index list.
        - Added more UTM-parameters to the default social query parameter list.
        - Renamed 'add_directory_index_regexp' to 'add_dir_index_regexp'.
        - Up requirements' versions.
        - Improved the documentation.

0.33    2015-12-12
        - Maintenance release:
            - Upped version numbers on required modules.
            - Changed license to MIT.
            - Updated dist.ini.

0.32    2015-10-29
        - ADDED: add_directory_index_regexp() can be used to add directory
          index regexps.
        - ADDED: add_social_query_param() can be used to add social query
          params.
        - Added a few tests.
        - Improved the documentation.

0.31    2015-10-28
        - Improved the documentation.

0.30    2015-10-28
        - MAJOR UPDATE: Lots of changes under the hood, but most of the
          functionality should behave as before, with exceptions mentioned
          below.
        - Improved the tests.
        - Improved the documentation, although the API hasn't changed that
          much.
        - CHANGED: The constructor now only accepts a URL, _or_ a path of
          a URL. This makes it possible to do normalizing on just the paths
          as well, but the 'base' argument might come back in the future.
        - REMOVED: remove_hostname_prefix() is no longer considered to be
          the job of this module. Instead, look into Domain::PublicSuffix
          for getting the "real" domain name from a URL.
        - REMOVED: do_all() because it's considered too harmful.

0.22    2015-09-09
        - Added correct 'remove_social_query_parameters' call from 'do_all'.

0.21    2015-03-02
        - When removing dot segments, '/.../' will now be treated as '/./'.

0.20    2014-12-12
        - Added url method, which is an alias for get_url.
        - Upped URI version requirement to 1.65.
        - Improved the documentation.

0.19    2014-06-03
        - Added remove_social_query_parameters method.
        - Improved the remove_empty_query_parameters method.
        - Improved the documentation.

0.18    2014-05-31
        - Improved support for older Perl versions.

0.17    2014-05-31
        - remove_fragment does NOT remove the fragment (#whatever) if they
          are part of the "logical" path, ie. "http://www.ex.com/#foo/bar"
          will NOT be changed, but "http://www.ex.com/#foo" will be translated
          to "http://www.ex.com/".
          Added tests for it.
        - remove_fragments was added, which behaves like the old
          remove_fragment, and removes anything after a # in the URL.
          Added tests for it.
        - Rewrote the remove_dot_segments method totally, and it no longer
          relies on the RFC 3986 algorithm, but something simpler or better,
          depending on how one chooses to look at it. (jamaica)
          Added tests for it.
        - Improved the documentation.

0.16    2013-07-21
        - Only try to sort query parameters if there are query parameters
          present.

0.15    2013-02-18
        - Improved error message when illegal URL and/or base is used.

0.14    2013-02-07
        - Added URL validity checking. (@Theo_vanHoesel).

0.13    2012-12-18
        - Upped the number of times remove_dot_segments can run before
          giving up.

0.12    2012-12-18
        - Fixed a bug in remove_dot_segments where the removal of multiple
          slashes were too aggressive.
        - Added more tests.

0.11    2012-12-13
        - Made it possible to modify the directory index regexps, so that the
          remove_directory_index method can be modified to your own fitting.
        - Removed old (commented) code.
        - Improved the documentation.

0.10    2012-11-02
        - No longer requires Moose.
        - Added more tests.
        - Improved the documentation.

0.09    2012-10-31
        - Fixed a bug where a cached URI object was returned from get_URI.
        - Improved the documentation.

0.08    2012-10-30
        - Added the remove_hostname_prefix method.
        - Exposed the get_URI method.
        - Added more tests.
        - Improved the documentation.

0.07    2012-10-30
        - Renamed several of the tests' filename.

lib/URL/Normalize.pm
package URL::Normalize;
use Moose;
use namespace::autoclean;

use URI qw();
use URI::QueryParam qw();

=head1 NAME

URL::Normalize - Normalize/optimize URLs.

=head1 VERSION

Version 0.44

=cut

our $VERSION = '0.44';

=head1 SYNOPSIS

    use URL::Normalize;

    my $normalizer = URL::Normalize->new( 'http://www.example.com/display?lang=en&article=fred' );

    # Normalize the URL
    $normalizer->make_canonical;
    $normalizer->remove_directory_index;
    $normalizer->remove_empty_query;

    # Get the normalized version back
    my $url = $normalizer->url;

=cut

=head1 DESCRIPTION

When writing a web crawler, for example, it's always very costly to check if a
URL has been fetched/seen when you have millions or billions of URLs in a
database.

This module can help you create a unique "ID" of a URL, which you can use as a
key in a key/value-store; the key is the normalized URL, whereas all the URLs
that refers to the normalized URL are part of the value (normally an array or
hash);

    'http://www.example.com/' = {
        'http://www.example.com:80/'        => 1,
        'http://www.example.com/index.html' => 1,
        'http://www.example.com/?'          => 1,
    }

Above, all the URLs inside the hash normalizes to the key if you run these
methods:

=over 4

=item * C<make_canonical>

=item * C<remove_directory_index>

=item * C<remove_empty_query>

=back

This is NOT a perfect solution.

If you normalize a URL using all the methods in this module, there is a high
probability that the URL will stop "working." This is merely a helper module
for those of you who wants to either normalize a URL using only a few of the
safer methods, and/or for those of you who wants to generate a possibly unique
"ID" from any given URL.

=head1 CONSTRUCTORS

=head2 new( $url )

Constructs a new URL::Normalize object:

    my $normalizer = URL::Normalize->new( 'http://www.example.com/some/path' );

You can also send in just the path:

    my $normalizer = URL::Normalize->new( '/some/path' );

The latter is NOT recommended, though, and hasn't been tested properly. You
should always give URL::Normalize an absolute URL by using L<URI>'s C<new_abs>.

=cut

=head1 METHODS

=cut

around BUILDARGS => sub {
    my $orig  = shift;
    my $class = shift;

    if ( @_ == 1 && !ref $_[0] ) {
        return $class->$orig( url => $_[0] );
    }
    else {
        return $class->$orig( @_ );
    }
};

=head2 url

Get the current URL, preferably after you have run one or more of the
normalization methods.

=cut

has 'url' => (
    isa      => 'Str',
    is       => 'ro',
    required => 1,
    writer   => '_set_url',
);

has 'dir_index_regexps' => (
    traits  => [ 'Array' ],
    isa     => 'ArrayRef[Str]',
    is      => 'rw',
    handles => {
        'add_dir_index_regexp' => 'push',
    },
    default => sub {
        [
            '/default\.aspx?',
            '/default\.html\.aspx?',
            '/default\.s?html?',
            '/home\.s?html?',
            '/index\.cgi',
            '/index\.html\.aspx?',
            '/index\.html\.php',
            '/index\.jsp',
            '/index\.php\d?',
            '/index\.pl',
            '/index\.s?html?',
            '/welcome\.s?html?',
        ];
    },
);

=head2 URI

Returns a L<URI> representation of the current URL.

=cut

has 'URI' => (
    isa => 'URI',
    is => 'ro',
    lazy => 1,
    default => sub {
        my $self = shift;

        my $URI = eval {
            URI->new( $self->url )->canonical;
        };

        if ( $@ ) {
            Carp::carp( "Failed to create a URI object from URL '" . $self->url . "'" );
        }

        return $URI;
    },
);

=head2 make_canonical

Just a shortcut for URI::URL->new->canonical->as_string, and involves the
following steps (at least):

=over 4

=item * Converts the scheme and host to lower case.

=item * Capitalizes letters in escape sequences.

=item * Decodes percent-encoded octets of unreserved characters.

=item * Removes the default port (port 80 for http).

=back

Example:

    my $normalizer = URL::Normalize->new(
        url => 'HTTP://www.example.com:80/%7Eusername/',
    );

    $normalizer->make_canonical;

    print $normalizer->url; # http://www.example.com/~username/

=cut

sub make_canonical {
    my $self = shift;

    if ( my $URI = $self->URI ) {
        $self->_set_url( $URI->canonical->as_string );
    }
    else {
        Carp::carp( "Can't make non-URI URLs canonical." );
    }
}

=head2 remove_dot_segments

The C<.>, C<..> and C<...> segments will be removed and "folded" (or
"flattened", if you prefer) from the URL.

This method does NOT follow the algorithm described in L<RFC 3986: Uniform
Resource Indentifier|http://tools.ietf.org/html/rfc3986>, but rather flattens
each path segment.

Also keep in mind that this method doesn't (because it can't) account for
symbolic links on the server side.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/../a/b/../c/./d.html',
    );

    $normalizer->remove_dot_segments;

    print $normalizer->url; # http://www.example.com/a/c/d.html

=cut

sub remove_dot_segments {
    my $self = shift;

    if ( my $URI = $self->URI ) {
        my $path = $URI->path;

        my @new_segments = ();

        foreach my $segment ( split('/', $path) ) {
            if ( $segment eq '.' || $segment eq '...' ) {
                next;
            }

            if ( $segment eq '..' ) {
                pop( @new_segments );
                next;
            }

            push( @new_segments, $segment );
        }

        my $new_path = join( '/', @new_segments );
        $new_path .= '/' if ( $new_path !~ m,/$, && $path =~ m,/$, );
        $new_path  = '/' . $new_path unless ( $new_path =~ m,^/, );

        $URI->path( $new_path );

        $self->_set_url( $URI->as_string );
    }
}

=head2 remove_directory_index

Removes well-known directory indexes, eg. C<index.html>, C<default.asp> etc.
This method is case-insensitive.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/index.cgi?foo=/',
    );

    $normalizer->remove_directory_index;

    print $normalizer->url; # http://www.example.com/?foo=/

The default regular expressions for matching a directory index are:

=over 4

=item * C<default\.aspx?>

=item * C<default\.html\.aspx?>

=item * C<default\.s?html?>

=item * C<home\.s?html?>

=item * C<index\.cgi>

=item * C<index\.html\.aspx?>

=item * C<index\.html\.php>

=item * C<index\.jsp>

=item * C<index\.php\d?>

=item * C<index\.pl>

=item * C<index\.s?html?>

=item * C<welcome\.s?html?>

=back

You can override these by sending in your own list of regular expressions
when creating the URL::Normalizer object:

    my $normalizer = URL::Normalize->new(
        url               => 'http://www.example.com/index.cgi?foo=/',
        dir_index_regexps => [ 'MyDirIndex\.html' ], # etc.
    );

You can also choose to add regular expressions after the URL::Normalize
object has been created:

    my $normalizer = URL::Normalize->new(
        url               => 'http://www.example.com/index.cgi?foo=/',
        dir_index_regexps => [ 'MyDirIndex\.html' ], # etc.
    );

    # ...

    $normalizer->add_directory_index_regexp( 'MyDirIndex\.html' );

=cut

sub remove_directory_index {
    my $self = shift;

    if ( my $URI = $self->URI ) {
        if ( my $path = $URI->path ) {
            my @compiled_regexps = map { qr/$_/i } @{$self->dir_index_regexps};

            foreach my $regex ( @compiled_regexps ) {
                $path =~ s,$regex,/,;
            }

            $URI->path($path);
        }

        $self->_set_url( $URI->as_string );
    }
}

=head2 sort_query_parameters

Sorts the URL's query parameters alphabetically.

Uppercased parameters will be lowercased DURING sorting, but the parameters
will be in the original case AFTER sorting. If there are multiple values for
one parameter, the key/value-pairs will be sorted as well.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/?b=2&c=3&a=0&A=1',
    );

    $normalizer->sort_query_parameters;

    print $normalizer->url; # http://www.example.com/?a=0&A=1&b=2&c=3

=cut

sub sort_query_parameters {
    my $self = shift;

    if ( my $URI = $self->URI ) {
        if ( $URI->as_string =~ m,\?, ) {
            my $query_hash     = $URI->query_form_hash || {};
            my $query_string   = '';
            my %new_query_hash = ();

            foreach my $key ( sort { lc($a) cmp lc($b) } keys %{$query_hash} ) {
                my $values = $query_hash->{ $key };
                unless ( ref $values ) {
                    $values = [ $values ];
                }

                foreach my $value ( @{$values} ) {
                    push( @{ $new_query_hash{lc($key)}->{$value} }, $key );
                }
            }

            foreach my $sort_key ( sort keys %new_query_hash ) {
                foreach my $value ( sort keys %{$new_query_hash{$sort_key}} ) {
                    foreach my $key ( @{$new_query_hash{$sort_key}->{$value}} ) {
                        $query_string .= $key . '=' . $value . '&';
                    }
                }
            }

            $query_string =~ s,&$,,;

            $URI->query( $query_string );
        }

        $self->_set_url( $URI->as_string );
    }
}

=head2 remove_duplicate_query_parameters

Removes duplicate query parameters, i.e. where the key/value combination is
identical with another key/value combination.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/?a=1&a=2&b=4&a=1&c=4',
    );

    $normalizer->remove_duplicate_query_parameters;

    print $normalizer->url; # http://www.example.com/?a=1&a=2&b=3&c=4

=cut

sub remove_duplicate_query_parameters {
    my $self = shift;

    if ( my $URI = $self->URI ) {
        my %seen      = ();
        my @new_query = ();

        foreach my $key ( $URI->query_param ) {
            my @values = $URI->query_param( $key );

            foreach my $value ( @values ) {
                unless ( $seen{$key}->{$value} ) {
                    push( @new_query, { key => $key, value => $value } );
                    $seen{$key}->{$value}++;
                }
            }
        }

        my $query_string = '';
        foreach ( @new_query ) {
            $query_string .= $_->{key} . '=' . $_->{value} . '&';
        }

        $query_string =~ s,&$,,;

        $URI->query( $query_string );

        $self->_set_url( $URI->as_string );
    }
}

=head2 remove_empty_query_parameters

Removes empty query parameters, i.e. where there are keys with no value. This
only removes BLANK values, not values considered to be no value, like zero (0).

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/?a=1&b=&c=3',
    );

    $normalizer->remove_empty_query_parameters;

    print $normalizer->url; # http://www.example.com/?a=1&c=3

=cut

sub remove_empty_query_parameters {
    my $self = shift;

    if ( my $URI = $self->URI ) {
        foreach my $key ( $URI->query_param ) {
            my @values = $URI->query_param( $key );

            $URI->query_param_delete( $key );

            foreach my $value ( @values ) {
                if ( defined $value && length $value ) {
                    $URI->query_param_append( $key, $value );
                }
            }
        }

        $self->_set_url( $URI->as_string );
    }
}

=head2 remove_empty_query

Removes empty query from the URL.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/foo?',
    );

    $normalizer->remove_empty_query;

    print $Normalize->url; # http://www.example.com/foo

=cut

sub remove_empty_query {
    my $self = shift;

    my $url = $self->url;
    $url =~ s,\?$,,;

    $self->_set_url( $url );
}

=head2 remove_fragment

Removes the fragment from the URL, but only if seems like they are at the end
of the URL.

For example C<http://www.example.com/#foo> will be translated to
C<http://www.example.com/>, but C<http://www.example.com/#foo/bar> will stay
the same.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/bar.html#section1',
    );

    $normalizer->remove_fragment;

    print $normalizer->url; # http://www.example.com/bar.html

=cut

sub remove_fragment {
    my $self = shift;

    my $url = $self->url;

    $url =~ s{#(?:/|[^?/]*)$}{};

    $self->_set_url( $url );
}

=head2 remove_fragments

Like C<remove_fragment>, but removes EVERYTHING after a C<#>.

=cut

sub remove_fragments {
    my $self = shift;

    my $url = $self->url;

    $url =~ s/#.*//;

    $self->_set_url( $url );
}

=head2 remove_duplicate_slashes

Remove duplicate slashes from the URL.

Example:

    my $normalizer = URL::Normalize->new(
        url => 'http://www.example.com/foo//bar.html',
    );

    $normalizer->remove_duplicate_slashes;

    print $normalizer->url; # http://www.example.com/foo/bar.html

=cut

sub remove_duplicate_slashes {
    my $self = shift;

    if ( my $URI = $self->URI ) {
        my $path = $URI->path;

        $path =~ s,/+,/,g;

        $URI->path( $path );

        $self->_set_url( $URI->as_string );
    }
}

=head2 remove_query_parameter

Convenience method for removing a specific parameter from the URL. If
the parameter is mentioned multiple times (?a=1&a=2), all occurences
will be removed.

=cut

sub remove_query_parameter {
    my $self  = shift;
    my $param = shift;

    return $self->remove_query_parameters( [$param] );
}

=head2 remove_query_parameters

Convenience method for removing multiple parameters from the URL. If the
parameters are mentioned multiple times (?a=1&a=2), all occurences will be
removed.

=cut

sub remove_query_parameters {
    my $self   = shift;
    my $params = shift || [];

    if ( my $URI = $self->URI ) {
        foreach my $param ( @{$params} ) {
            $URI->query_param_delete( $param );
        }

        $self->_set_url( $URI->as_string );
    }
}

=head1 SEE ALSO

=over 4

=item * L<URI::Normalize>

=item * L<URI>

=item * L<URI::URL>

=item * L<URI::QueryParam>

=item * L<RFC 3986: Uniform Resource Indentifier|http://tools.ietf.org/html/rfc3986>

=item * L<Wikipedia: URL normalization|http://en.wikipedia.org/wiki/URL_normalization>

=back

=head1 AUTHOR

Tore Aursand, C<< <toreau at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to the web interface at L<https://github.com/toreau/URL-Normalize/issues>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc URL::Normalize

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/URL-Normalize>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/URL-Normalize>

=item * Search CPAN

L<http://search.cpan.org/dist/URL-Normalize/>

=back

=head1 LICENSE AND COPYRIGHT

The MIT License (MIT)

Copyright (c) 2012-2021 Tore Aursand

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

=cut

__PACKAGE__->meta->make_immutable;

1; # End of URL::Normalize
